### DISCUSSION TITLE
how to disable `invalid_key_behavior`?

### DISCUSSION THREAD

**Discussion by chrisgrieser** [kylechui/nvim-surround#213] [Q&A]
I find myself more often accidentally deleting something than actually using the `invalid_key_behavior` (= automatic surrounds). Is there any way to disable that? Can't find anything in the docs on that

**Comment by chrisgrieser** [comment]
hmm, so I found this, which at least results in an error instead instead of adding/deleting characters unintentionally.

```lua
require("nvim-surround").setup {
  surrounds = {
	invalid_key_behavior = {
		add = {},
		find = nil,
		delete = nil,
		change = {},
    },
  }
}
```

**Reply by kylechui** [reply]
Can you please clarify your question by providing concrete examples of what you would like to happen? Also perhaps read the help menu on `invalid_key_behavior` and see if it answers your question.

**Reply by chrisgrieser** [reply]
Thanks you for the quick reply! As I said, unfortunately I was not able to find anything in the help regarding this.

What I want is that when I use an invalid key, nothing should happen. (Having *nothing* set up for `v` as alias)
```lua
-- start
foo|bar

-- result after using typing `ysiwv`
vfoo|barv

-- what I want to happen (= nothing should happen)
foo|bar
```

Like, I want to disable the behavior that an invalid key does *not* surround the text object with anything. Reason being that more often than being helpful, I end up accidentally adding/deleting things.

**Reply by kylechui** [reply]
Ah I see! To accomplish this, set the `invalid_key_behavior` keys to effectively do nothing, i.e. `add` can be set to `{ "", "" }`. If you need help with the other keys, let me know; extra information on what return types the keys expect can be found in the help menu for nvim-surround.

**Reply by chrisgrieser** [reply]
Ah, I see. So I got it working for `add` and `delete`:
```lua
invalid_key_behavior = {
	add = function(_) return { { "" }, { "" } } end,
	find = function(char) return config.get_selection { pattern = vim.pesc(char) .. ".-" .. vim.pesc(char) } end,
	delete = function(_)
		return config.get_selections {
			char = "",
			pattern = "^()().-()()$",
		}
	end,
},
```

But I cannot seems to figure out what I have to add for the change to do nothing. Emptying the string just results in `cs` deleting the surrounds, instead of leaving them untouched. Any pointers?

---

I think it might also make sense to simply add an simple boolean option to disable this; having to figure out how the syntax works just to disable an extra feature the plugin provides feels somewhat unnecessarily complicated 🥴

**Reply by kylechui** [reply]
That sounds like a reasonable idea; I'll write that up right now as a way for people to just set things to `false` in order to "disable" the plugin in a sensible way.

**Reply by kylechui** [reply]
@chrisgrieser Fair warning: I haven't looked at this spaghetti code in a while, and only now is it starting to make sense again. It'll probably be a bit before I'm able to issue an update, since it seems that I should rewrite some code to make it more logical.

**Reply by kylechui** [reply]
@chrisgrieser Do you mind testing out the latest commit on `main` for me? Try any sort of configuration that you feel like "is reasonable", and let me know how it goes. If there's anything that's unexpected, feel free to tag me with the corresponding configuration that caused the issue, and what happened vs. what was expected.

**Reply by chrisgrieser** [reply]
@kylechui thanks! tried the latest commit. I tried this config:
```lua
invalid_key_behavior = {
	add = false,
	find = false,
	delete = false,
	change = false,
},
```

is this the indented one? anyway, while they do not change the next in an unintended manner, I get errors with all three, `cs`, `ys`, and `ds`
- `cs` and `ds`: invalid key provided for `:h nvim-surround.config.get_selection()`
- `ys` leaves the text object selected and gives this error

![CleanShot 2023-03-02 at 19 12 46](https://user-images.githubusercontent.com/73286100/222516007-79998bcf-682d-46a8-a39c-7a33d35779ea.png)

**Reply by kylechui** [reply]
Hmm... adding that into the `surrounds` config works perfectly fine for me. Seeing that you're using `lazy.nvim`, do you mind trying to "unlazily" load my plugin? I have a sneaking suspicion that some commands aren't being fired in the correct order due to lazy loading.

**Reply by chrisgrieser** [reply]
tired un-lazy-loading, it does not seem to make a difference

**Reply by kylechui** [reply]
Have you updated the plugin? I checked some of the line numbers in the error message and they seem to not be lining up properly.

**Reply by chrisgrieser** [reply]
yeah, I am on the latest commit of the main branch:
![CleanShot 2023-03-03 at 00 56 54](https://user-images.githubusercontent.com/73286100/222593630-b1ecd153-eba0-4c57-973b-a3f6cea40d9f.png)

**Reply by kylechui** [reply]
Could you try:
* Implement a minimal config that replicates the issue
* Switch to packer and see if the issue persists

**Reply by chrisgrieser** [reply]
Ok that's weird. After disabling my complete nvim-surround config (except the invalid-key-behavior), it suddenly worked. and when I re-enabled the config, it kept working 😆 

My guess is that this is due to lazy's weird caching mechanism running the background or something?

Anyway, issue seems resolved – thanks for the assist! 😊
